Centralize rainbird config and add binary sensor platform

konikvranik opened the pull request:
## Breaking Change:
The configuration has been changed and all configuration is now done under the `rainbird:` key in configuration.yaml. Configuration of rainbird switches is moved from platform config section to main `rainbird:` config under `zones:` key. Please read the updated documentation. 

## Description:
- [x] Bump pyrainbird to 0.4.1
- Cache states of sprinklers for efficient communication to controller
- fixed rainsensor state
- added rainsensor binary sensor

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io):** home-assistant/home-assistant.io#10326

## Example entry for `configuration.yaml`:
````yaml
rainbird:
  host: IP_ADDRESS_OF_MODULE
  password: YOUR_PASSWORD
  trigger_time: 6
  zones:
    1:
      trigger_time: 6
````
## Checklist:
  - [x] The code change is tested and works locally.
  - [x]  Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

konikvranik committed:
Update pyrainbird to version 0.2.0 to fix zone number issue:

- home-assistant/home-assistant/issues/24519
- jbarrancos/pyrainbird/issues/5
- https://community.home-assistant.io/t/rainbird-zone-switches-5-8-dont-correspond/104705

konikvranik committed:
requirements_all.txt regenerated

konikvranik committed:
code formatting

konikvranik committed:
pyrainbird version 0.3.0

konikvranik committed:
zone id

konikvranik committed:
rainsensor return state

konikvranik committed:
updating rainsensor

konikvranik committed:
new version of pyrainbird

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,74 @@
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)): vol.All(
```
We don't allow monitored conditions in config anymore.

https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md

konikvranik replied:
there is single api request for every rainbird sensor state.
Not all are implemented yet, but according the spec CONF_MONITORED_CONDITIONS should be used in this case

MartinHjelmare replied:
It's allowed to have monitored conditions if the amount of requests is a problem. We still prefer to not have this option if the integration and api can cope.

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,74 @@
+        self._controller = controller
+        self._name = SENSOR_TYPES[self._sensor_type][0]
+        self._icon = SENSOR_TYPES[self._sensor_type][2]
+        self._unit_of_measurement = SENSOR_TYPES[self._sensor_type][1]
```
Binary sensors don't have a unit of measurement.

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,74 @@
+        self._state = None
+
+    @property
+    def state(self):
```
Don't overwrite the state property. Implement `is_on` property.

konikvranik committed:
binary sensor state

konikvranik commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,74 @@
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_MONITORED_CONDITIONS
+from homeassistant.helpers.entity import Entity
+from pyrainbird import RainbirdController
```
whats wrong with imports?

MartinHjelmare replied:
Please group the imports in standard library, 3rd party and homeassistant import groups.

https://www.python.org/dev/peps/pep-0008/#imports

konikvranik replied:
so homeassistant imports should be the last?

konikvranik replied:
I use pycharm formatter which formats it the way it is commited
is it configurable somewhat?

MartinHjelmare replied:
The `isort` tool will sort the imports automatically. Most advanced editors have isort plugins.

konikvranik replied:
No luck with isort for pycharm :-/ But commandline isort did the work

konikvranik committed:
quiet in check format

konikvranik committed:
is_on instead of state for binary_sensor

konikvranik committed:
no unit of measurement for binary sensor

konikvranik committed:
no monitored conditions config

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -15,21 +10,13 @@
 
     sensors = []
-    for sensor_type in config.get(CONF_MONITORED_CONDITIONS):
+    for sensor_type in map(lambda k, v: k, SENSOR_TYPES):
```
```py
for sensor_type in SENSOR_TYPES:
```

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,55 @@
+
+    @property
+    def is_on(self):
+        self._state = self._controller.get_rain_sensor_state()
```
This is already done in the `update` method.

konikvranik committed:
get keys of dict directly

konikvranik committed:
removed redundant update of state

konikvranik committed:
simplified switch

konikvranik committed:
right states for switch

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -16,7 +16,7 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
     sensors = []
-    for sensor_type in map(lambda k, v: k, SENSOR_TYPES):
+    for sensor_type in SENSOR_TYPES.keys():
```
Iterating a dictionary iterates the keys by default. Please remove `.keys()`.

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,54 @@
+    def update(self):
+        """Get the latest data and updates the states."""
+        _LOGGER.debug("Updating sensor: %s", self._name)
+        if self._sensor_type == "rainsensor":
```
Do we need this check? There's only one sensor type.

konikvranik replied:
There are another values we will get from rainbird device. Not all are implemented yet. The rainsensor is most important, raindelay is the next.

MartinHjelmare replied:
Please break out the sensor type strings into constants, since we use these in multiple places.

konikvranik replied:
yep

konikvranik committed:
raindelay sensor

konikvranik committed:
raindelay sensor

konikvranik committed:
binary sensor state

konikvranik committed:
binary sensor state

konikvranik committed:
reorganized imports

konikvranik committed:
doc on public method

konikvranik committed:
reformatted

peternijssen committed:
add irrigation service to rain bird, which allows you to set the duration

peternijssen committed:
rebased on konikvranik and solved some feedback

MartinHjelmare reviewed (commented):
Since this PR already is a breaking change, I suggest we also change to use `discovery.load_platform` to load the platforms from the component, instead of using platform config sections. This is our new standard to collect all config under the integration key in the config.

This will make it easier to continue expanding this integration.

konikvranik committed:
add irrigation service to rain bird

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -45,6 +56,22 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            if isinstance(device, RainBirdSwitch) and device.entity_id == entity_id:
+                device.start_irrigation(duration)
+
+    hass.services.async_register(
```
We're in sync context so have to use the sync api. Use `hass.services.register`.

konikvranik replied:
yes

MartinHjelmare commented on tests/components/python_script/test_init.py:
```diff
@@ -64,7 +64,7 @@ def test_execute_with_data(hass, caplog):
     """
 
-    hass.async_add_job(execute, hass, "test.py", source, {"name": "paulus"})
+    hass.async_add_job(execute, hass, "integration_test.py", source, {"name": "paulus"})
```
Please don't change these files.

konikvranik replied:
this was an unwanted commit. reverted back.

MartinHjelmare reviewed (commented):
Please revert unrelated changes.

konikvranik committed:
sensor types to constants

konikvranik committed:
synchronized register service

konikvranik committed:
patform discovery

konikvranik reviewed (commented):
> Since this PR already is a breaking change, I suggest we also change to use `discovery.load_platform` to load the platforms from the component, instead of using platform config sections. This is our new standard to collect all config under the integration key in the config.
> 
> This will make it easier to continue expanding this integration.

Please look at d6dd9162ca755b47473d3f063382b6a730c64098  if that's the way you intended

konikvranik committed:
binary sensor as wrapper to sensor

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -41,10 +53,13 @@ def setup(hass, config):
+            hass,
+            platform,
+            DOMAIN,
+            discovered={RAINBIRD_CONTROLLER: controller},
```
Only serializable info can be sent. Store the controller in hass.data instead like before.

Pass an empty dict as discovery_info.

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/sensor.py:
```diff
@@ -1,44 +1,53 @@
+        )
+        devices += [sensor]
+
+        discovery.load_platform(
```
Remove this.

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,48 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up a Rain Bird sensor."""
+    if discovery_info is None or not PARENT_SENSOR in discovery_info:
```
```py
if discovery_info is None:
    return
```

konikvranik replied:
done

konikvranik committed:
version 0.4.0

konikvranik committed:
new config approach

konikvranik committed:
sensors cleanup

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -2,60 +2,100 @@
+
+SERVICE_START_IRRIGATION = "start_irrigation"
+
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Remove the platform schemas. They will not be used when loading the integration via the component.

konikvranik replied:
platform schema removed.

konikvranik committed:
bypass if no zones found

MartinHjelmare reviewed (commented):
Please don't add more features to this PR after this.

konikvranik committed:
platform schema removed

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,79 @@
+    }
+)
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.Any({cv.string: CONTROLLER_SCHEMA}, CONTROLLER_SCHEMA))},
```
The controller id doesn't need to be a config item. We can just use some arbitrary integer as index. Use `cv.ensure_list` and make it a list of controller schemas.
```py
CONFIG_SCHEMA = vol.Schema(
    {DOMAIN: vol.Schema(vol.All(cv.ensure_list, [CONTROLLER_SCHEMA]))},

konikvranik replied:
refactored

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,79 @@
 
-    from pyrainbird import RainbirdController
+    hass.data[DATA_RAINBIRD] = dict()
+    if CONF_HOST in config[DOMAIN] and CONF_PASSWORD in config[DOMAIN]:
```
This case can be removed with the above change of the config schema.

konikvranik replied:
refactored

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,79 @@
-        _LOGGER.error("Error getting state. Possible configuration issues")
-        return False
+def _setup_controller(config, controller_id, hass):
+    from pyrainbird import RainbirdController
```
Please move 3rd party imports to the top of the module.

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,79 @@
 
-    hass.data[DATA_RAINBIRD] = controller
-    return True
+    server = config.get(CONF_HOST)
```
```suggestion
    server = config[CONF_HOST]
```

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,79 @@
-    hass.data[DATA_RAINBIRD] = controller
-    return True
+    server = config.get(CONF_HOST)
+    password = config.get(CONF_PASSWORD)
```
```suggestion
    password = config[CONF_PASSWORD]
```

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,79 @@
-    if initial_status and initial_status["type"] != "CurrentStationsActiveResponse":
-        _LOGGER.error("Error getting state. Possible configuration issues")
-        return False
+def _setup_controller(config, controller_id, hass):
```
Start with `hass` parameter.
```suggestion
def _setup_controller(hass, config, controller_id):
```

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,79 @@
+    controller = RainbirdController(server, password)
+    hass.data[DATA_RAINBIRD][controller_id] = controller
+    _LOGGER.debug("Rain Bird Controller %s set to: %s", controller_id, server)
+    for platform in [switch.DOMAIN, sensor.DOMAIN, binary_sensor.DOMAIN]:
```
Break out the list of supported platforms to a module level constant.

MartinHjelmare commented on homeassistant/components/rainbird/binary_sensor.py:
```diff
@@ -0,0 +1,64 @@
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up a Rain Bird sensor."""
+    if discovery_info is None:
+        return False
```
```suggestion
        return
```

MartinHjelmare commented on homeassistant/components/rainbird/sensor.py:
```diff
@@ -1,44 +1,39 @@
 
-# sensor_type [ description, unit, icon ]
-SENSOR_TYPES = {"rainsensor": ["Rainsensor", None, "mdi:water"]}
+PARENT_SENSOR = "parent_sensor"
```
Where is this used?

konikvranik replied:
Not used now. Was for pairing binary_sensor to sensor. Removing.

MartinHjelmare commented on homeassistant/components/rainbird/sensor.py:
```diff
@@ -1,44 +1,39 @@
-    for sensor_type in config.get(CONF_MONITORED_CONDITIONS):
-        sensors.append(RainBirdSensor(controller, sensor_type))
+    if discovery_info is None:
+        return False
```
```suggestion
        return
```

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -4,58 +4,83 @@
-    controller = hass.data[DATA_RAINBIRD]
 
+    if discovery_info is None:
+        return False
```
```suggestion
        return
```

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -4,58 +4,83 @@
+    ]
+    available_stations: AvailableStations = controller.get_available_stations()
+    if not (available_stations and available_stations.stations):
+        return False
```
```suggestion
        return
```

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -4,58 +4,83 @@
     add_entities(devices, True)
 
+    def _start_irrigation(service):
+        entity_id = service.data.get(ATTR_ENTITY_ID)
```
```suggestion
        entity_id = service.data[ATTR_ENTITY_ID]
```

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -4,58 +4,83 @@
 
+    def _start_irrigation(service):
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        duration = service.data.get(ATTR_DURATION)
```
```suggestion
        duration = service.data[ATTR_DURATION]
```

konikvranik replied:
done

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -4,58 +4,83 @@
     """Representation of a Rain Bird switch."""
 
-    def __init__(self, rb, dev, dev_id):
+    def __init__(self, controller: RainbirdController, zone, time, name=None):
```
We don't need to make name optional.
```suggestion
    def __init__(self, controller: RainbirdController, zone, time, name):
```

konikvranik replied:
done

konikvranik committed:
Change config schema to list of controllers

some small code improvements as suggested in CR:
 - dictionary acces by []
 - just return instead of return False
 - import order
 - no optional parameter name

konikvranik commented:
There has to be upgrade to pyrainbird 0.4.1. Testing with HA I found bug there.

konikvranik reviewed (commented):
There has to be upgrade to pyrainbird 0.4.1. Testing with HA I found bug there

konikvranik committed:
some small code improvements as suggested in CR:
 - supported platforms in constant
 - just return instead of return False
 - removed unused constant

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,82 @@
+    }
+)
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.Any(cv.ensure_list(CONTROLLER_SCHEMA), CONTROLLER_SCHEMA))},
```
```suggestion
    {DOMAIN: vol.Schema(vol.All(cv.ensure_list, [CONTROLLER_SCHEMA]))},
```

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,82 @@
 
-    from pyrainbird import RainbirdController
+    hass.data[DATA_RAINBIRD] = list()
+    if isinstance(config[DOMAIN], list):
```
Remove all the checks, and just iterate the list of controllers. The config validation will make sure it's always a list.

konikvranik replied:
ok

MartinHjelmare replied:
Just to be clear, with the updated config, the user can provide a single controller and a list of controllers. The config validation will turn both cases into a list of controllers. That's what `cv.ensure_list` does.

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -3,40 +3,82 @@
-    password = conf.get(CONF_PASSWORD)
 
-    from pyrainbird import RainbirdController
+    hass.data[DATA_RAINBIRD] = list()
```
```suggestion
    hass.data[DATA_RAINBIRD] = []
```

konikvranik committed:
No single controller configuration

Co-Authored-By: Martin Hjelmare <marhje52@kth.se>

MartinHjelmare commented:
Make sure to bump the library versions here too. They are still at 0.4.0.

konikvranik commented:
Yes, I'm waiting for publishing it. I'm not owner of pyraindbird library so I have wait until @jbarrancos publishes it to pypi.
Then I will update it.

konikvranik committed:
pyrainbird 0.4.1

konikvranik committed:
individual switch configuration

konikvranik commented:
Upgraded to pyrainbird 0.4.1, tested on local instance and all seems to work flawlessly

konikvranik commented:
@jbarrancos May I ask you how does `scan_interval` works?
Home assistant core finds out if component has this flag in config and in that case it sets proper update interval?
Or how is update scheduled? I can not find docs which explain it.

konikvranik committed:
imports order

konikvranik committed:
generate default name out of entity

jbarrancos commented:
Yes, correct. It polls the “sensor” to know it’s state(s).

From: Petr Vraník <notifications@github.com>
Sent: Tuesday, 10 September 2019 10:40
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Jordy Barrancos <jordy@fusion-ict.nl>; Mention <mention@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Centralize rainbird config and add binary sensor platform (#26393)


@jbarrancos<https://github.com/jbarrancos> May I ask you how does scan_interval works?
Home assistant core finds out if component has this flag in config and in that case it sets proper update interval?
Or how is update scheduled? I can not find docs which explain it.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/26393?email_source=notifications&email_token=AHO4AOEYYUUO5V65NNG2IQ3QI5MMZA5CNFSM4ITJFVAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6KKFIQ#issuecomment-529834658>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AHO4AOE5SSD5T7GZSLGS5TDQI5MMZANCNFSM4ITJFVAA>.

MartinHjelmare commented:
Please add a paragraph in the PR description about the breaking change. Describe briefly what changed and what the user needs to do to cope with the breaking change.

konikvranik committed:
trigger time required for controller

konikvranik committed:
incorporated CR remarks:
- constant fo rzones
- removed SCAN_INTERVAL
- detection of success on initialization
- removed underscore
- refactored if/else
- empty line on end of file
- hass as first parameter

konikvranik committed:
import of library on top

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -1,42 +1,88 @@
+                hass_config=config,
+            )
+        success = True
+    except Exception as e:
```
Please only wrap the call that can error with try... except.

We should try to except specific exceptions. What error do we expect?

konikvranik replied:
We don't want to initialize platform when controller is unreachable. That's why also loop is inside the exception.

There is no specific exception. Just `Exception`

MartinHjelmare replied:
Please change like this:
```py
def _setup_controller(hass, controller_config, config):
    """Set up a controller."""
    server = controller_config[CONF_HOST]
    password = controller_config[CONF_PASSWORD]
    controller = RainbirdController(server, password)
    position = len(hass.data[DATA_RAINBIRD])
    try:
        controller.get_serial_number()
    except Exception as exc:
        _LOGGER.error("Unable to setup controller: %s", exc)
        return False
    hass.data[DATA_RAINBIRD].append(controller)
    _LOGGER.debug("Rain Bird Controller %d set to: %s", position, server)
    for platform in SUPPORTED_PLATFORMS:
        discovery.load_platform(
            hass, platform, DOMAIN, {RAINBIRD_CONTROLLER: position, **controller_config}, config,
        )
    return True
```

The library should preferably be updated to raise library specific exceptions. It's not good to raise general exceptions like `Exception` directly. Create a library base exception that inherits from `Exception`. Then create further library exceptions as needed that all inherit from the base library exception.

konikvranik replied:
done

konikvranik committed:
refactored

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -1,42 +1,87 @@
+    hass.data[DATA_RAINBIRD] = []
+    success = False
+    for controller_config in config[DOMAIN]:
+        success = success or _setup_controller(hass, controller_config)
```
```suggestion
        success = success or _setup_controller(hass, controller_config, config)
```

konikvranik committed:
Update homeassistant/components/rainbird/__init__.py

Co-Authored-By: Martin Hjelmare <marhje52@kth.se>

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -1,42 +1,87 @@
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_TRIGGER_TIME): cv.string,
```
We should implement a schema similar to the manual alarm_control_panel:
https://github.com/home-assistant/home-assistant/blob/770eeaf82f4800127e977f980f5438d0fa8edbc9/homeassistant/components/manual/alarm_control_panel.py#L109-L111

konikvranik replied:
this way?

MartinHjelmare commented on homeassistant/components/rainbird/switch.py:
```diff
@@ -2,61 +2,96 @@
+                CONF_TRIGGER_TIME, discovery_info.get(CONF_TRIGGER_TIME, 0)
+            )
+            name = zone_config.get(CONF_FRIENDLY_NAME)
+            if not time:
```
Instead of checking here we should validate already in the config schema and set default for each controller based on the main setting. See the manual alarm_control_panel schema for example:
https://github.com/home-assistant/home-assistant/blob/770eeaf82f4800127e977f980f5438d0fa8edbc9/homeassistant/components/manual/alarm_control_panel.py#L63-L75

konikvranik replied:
this way?

konikvranik committed:
validate time and set defaults

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -32,17 +33,31 @@
+            vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+            vol.Optional(CONF_TRIGGER_TIME): TRIGGER_TIME_SCHEMA,
+        },
+        _validator,
```
The validator should be applied in the `CONTROLLER_SCHEMA` after the dict with all items.

konikvranik committed:
set defaults on right place

konikvranik committed:
pylint bypass

konikvranik committed:
iterate over values

MartinHjelmare commented on homeassistant/components/rainbird/manifest.json:
```diff
@@ -3,7 +3,7 @@
+    "pyrainbird==0.4.1"
   ],
   "dependencies": [],
   "codeowners": []
```
Please add yourself as codeowner.

MartinHjelmare reviewed (approved):
Good! Can be merged when final comment is addressed and build passes.

konikvranik committed:
codeowner

konikvranik committed:
reverted changes:

* irrigation time just as positive integer. Making it complex does make
sense
* zone edfaults fullfiled at runtime. There is no information about
available zones in configuration time.

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -1,42 +1,89 @@
+    SENSOR_TYPE_RAINDELAY: ["Raindelay", None, "mdi:water-off"],
+}
+
+TRIGGER_TIME_SCHEMA = cv.positive_int
```
Why did we change?

konikvranik replied:
That doesn't work. The complex type was not serializable and allowed several interpretations of time what we don't need here. If we use that construct, than we need to detect type, parse it and finally transfer it to int representing minutes of irrigation.
Which is what we want to set up: minutes of irrigation.
So there will be useless overhead to convert minutes from config through some unwanted type to minutes.
So I reverted to previous functional state.

MartinHjelmare replied:
It's not useless. We want the time options to use the same validator. If we need an integer in the end, we e convert to an integer in the end of the validation.

Please change again and add the last integer validator

konikvranik replied:
here it is

konikvranik committed:
codeowners updated

konikvranik committed:
accept timedelta in irrigation time

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -1,42 +1,96 @@
+}
+
+TRIGGER_TIME_SCHEMA = vol.Any(
+    cv.positive_int,
```
It's already valid to pass integers as seconds in timedelta, in `cv.time_period`.

konikvranik replied:
Yes, but as seconds. Common irrigation time is between 1 to 30 minutes. I do not se the reason to force users multiply minutes by 60 mynually when they want to se minutes of irrigation.

MartinHjelmare replied:
They can set minutes by using the minutes interface in a time dict or a time string.

We use the timedelta interface all over home assistant to specify time interval. Keeping a consistent interface helps users.

MartinHjelmare replied:
Please remove this.

MartinHjelmare reviewed (commented):
Please also set the defaults as before.

konikvranik committed:
simplified time calculation

MartinHjelmare commented:
Please set the defaults as before too.

konikvranik commented:
I'm sorry, which defaults do you mean?

MartinHjelmare commented:
The defaults we removed here:
https://github.com/home-assistant/home-assistant/pull/26393/commits/60ae83aecc4e871b11a46b66d43efb8ba8b59473#diff-d575fd59bad63f44b30cd5e8ef6286baL39-L44

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -1,42 +1,92 @@
+
+TRIGGER_TIME_SCHEMA = vol.Any(
+    cv.positive_int,
+    vol.All(cv.time_period, cv.positive_timedelta, lambda td: td.total_seconds // 60),
```
We need to call `td.total_seconds` not just access it.

konikvranik commented:
If you mean _set_defaults to set irrigation times for zones in configuration time, it does not make sense. Zones are not known in configuration time, so could not be set. They are discovered from controller in time of platform setup.

konikvranik committed:
call total_seconds

MartinHjelmare commented:
We have a zone config schema. We should set the default trigger time in the zone schema.

konikvranik commented:
We can set values only for zones we want to overwrite some config. We can not overwrite irrigation time for zones that we don't have in configuration. So it is nonsense to substitute default twice. In config ant then in runtime. That will be code duplication. What will be that for?

MartinHjelmare commented:
Ok. Address the last comment and we're ready.

konikvranik committed:
irrigation time as seconds.

MartinHjelmare commented on homeassistant/components/rainbird/__init__.py:
```diff
@@ -1,42 +1,93 @@
+    SENSOR_TYPE_RAINDELAY: ["Raindelay", None, "mdi:water-off"],
+}
+
+TRIGGER_TIME_SCHEMA = vol.Any(
```
We don't need `vol.Any` anymore.

konikvranik committed:
simplified schema

MartinHjelmare reviewed (approved):
Good!

MartinHjelmare commented:
Can be merged when build passes.
